### PR TITLE
[MRESOLVER-329] More robust IO in DefaultTrackingFileManager

### PR DESCRIPTION
**Pull request by cstamas**
Make the code more robust to copy with concurrently accessed files and hopefully with Windows quirks re atomic move.

---

https://issues.apache.org/jira/browse/MRESOLVER-329

### CODE REVIEW

File: maven-resolver-util/src/main/java/org/eclipse/aether/util/FileUtils.java [hunk 1]
```diff
@@ -105,7 +106,16 @@ public Path getPath() {
 
             @Override
             public void move() throws IOException {
-                Files.move(tempFile, file, StandardCopyOption.ATOMIC_MOVE);
+                try {
+                    Files.move(tempFile, file, StandardCopyOption.ATOMIC_MOVE);
```
Comment (laeubi) [inline maven-resolver-util/src/main/java/org/eclipse/aether/util/FileUtils.java:110]: As mentioned in the other PR this is likely not what one wants. If the (atomic) move do not succeeds, there is no guarantee that the file has not changed (deleted, modified, ...) so retry the move will simply ignore this and overwrite everything. So callers of the move operation must retry the full operation.

Comment (cstamas) [reply maven-resolver-util/src/main/java/org/eclipse/aether/util/FileUtils.java:110]: I might be wrong, but this is what Jenkins does as well: tries atomic, and if fails, the fallback is plain copy+rm... Naturally, this will need some triage, to see is it working at all.

**Comment by michael-o** [comment]
I'd like better understand whether the usecases file w/o temp files. I want to exclude our locks to be faulty here.

File: maven-resolver-util/src/main/java/org/eclipse/aether/util/FileUtils.java [hunk 2]
```diff
@@ -105,7 +106,16 @@ public Path getPath() {
...
+                } catch (AccessDeniedException e) {
+                    // MRESOLVER-329: fallback to plain copy+rm, this usually happens on Windows
```
Comment (michael-o) [inline maven-resolver-util/src/main/java/org/eclipse/aether/util/FileUtils.java:112]: MRESOLVER-325 and not 329

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultTrackingFileManager.java [hunk 3]
```diff
@@ -75,6 +76,8 @@ public Properties update(File file, Map<String, String> updates) {
             if (Files.isReadable(filePath)) {
                 try (InputStream stream = Files.newInputStream(filePath)) {
                     props.load(stream);
+                } catch (NoSuchFileException e) {
+                    // MRESOLVER-329: ignore; in case of concurrent r/w Files.isReadable may return true
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultTrackingFileManager.java:80]: Please move this to another PR, both issues aren't necessarily related.

**Review by michael-o — CHANGES_REQUESTED** [review]

**Comment by michael-o** [comment]
Does not work and suffers from the chicken and egg problem described in here: https://issues.apache.org/jira/browse/MRESOLVER-325?focusedCommentId=17693545&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17693545

**Comment by cstamas** [comment]
Superseded by https://github.com/apache/maven-resolver/pull/260

**Comment by jira-importer** [comment]
Resolve #1011
